records_mover/url/optimizer/__init__.py

```diff
@@ -0,0 +1,349 @@
+import json
+import time
+import datetime
+from contextlib import contextmanager
+from records_mover.url.base import BaseDirectoryUrl
+import logging
+from typing import Iterator, Tuple, Union, Optional, TYPE_CHECKING
+if TYPE_CHECKING:
+    from googleapiclient.discovery import _TransferJobConfig
+    import google.auth.credentials
+    from records_mover.url.gcs.gcs_directory_url import GCSDirectoryUrl
+    from records_mover.url.s3.s3_directory_url import S3DirectoryUrl
+    from records_mover.url.filesystem import FilesystemDirectoryUrl
+
+
+logger = logging.getLogger(__name__)
+
+
+class CopyOptimizer:
+    """When Records Mover needs to move data between different URL
+    schemes, the default way (a single-threaded downloading of bytes
+    from the source and then reuploading of bytes to the target) can
+    be very slow for large directories.
+
+    This class can use faster means (ideally a cloud-based service
+    that can copy things at scale, or at least vendor-optimized
+    multi-threaded upload/download code) when they are available.
+
+    This depends on the specifics of the copy, and as such return
+    values must be checked to determine if the copy was able to be
+    performed.
+    """
+
+    def copy(self,
+             loc: BaseDirectoryUrl,
+             other_loc: BaseDirectoryUrl) -> bool:
+        """Copy from one directory to another using optimized means.
+
+        :return: True if the copy was performed, or False if no
+        optimized means were possible."""
+        if loc.scheme == 's3' and other_loc.scheme == 'file':
+            from records_mover.url.filesystem import FilesystemDirectoryUrl
+            from records_mover.url.s3.s3_directory_url import S3DirectoryUrl
+
+            assert isinstance(loc, S3DirectoryUrl)
+            assert isinstance(other_loc, FilesystemDirectoryUrl)
+            return self._copy_via_awscli(loc, other_loc)
+        elif loc.scheme == 's3' and other_loc.scheme == 'gs':
+            from records_mover.url.gcs.gcs_directory_url import GCSDirectoryUrl
+            from records_mover.url.s3.s3_directory_url import S3DirectoryUrl
+
+            assert isinstance(loc, S3DirectoryUrl)
+            assert isinstance(other_loc, GCSDirectoryUrl)
+            return self._copy_via_gcp_data_transfer(loc, other_loc)
+        else:
+            logger.info(f"No strategy to optimize copy from {loc} to {other_loc}")
+            return False
+
+    def _copy_via_awscli(self,
+                         loc: 'S3DirectoryUrl',
+                         other_loc: 'FilesystemDirectoryUrl') -> bool:
+        from .awscli import aws_cli
+
+        # TODO: Can I pass in creds?
+        aws_cli('s3', 'sync', loc.url, other_loc.local_file_path)
+        return True
+
+    def _gcp_data_transfer_min_bytes_to_bother(self) -> int:
+        # TODO Add config pulling
```

Comment:
Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker